src/mesh/voxel_cloud.rs

```diff
@@ -0,0 +1,499 @@
+            (max_x_index - min_x_index + 2.0) as u32,
+            (max_y_index - min_y_index + 2.0) as u32,
+            (max_z_index - min_z_index + 2.0) as u32,
+        );
+
+        // Going to populate the mesh with points as dense as the smallest voxel dimension.
+        let shortest_voxel_dimension = voxel_dimensions
+            .x
+            .min(voxel_dimensions.y.min(voxel_dimensions.z));
+
+        let mut voxel_cloud = VoxelCloud::new(&block_start, &block_dimensions, voxel_dimensions);
+
+        // Iterate through mesh faces.
+        for face in mesh.faces() {
+            match face {
+                Face::Triangle(triangle_face) => {
+                    let point_a = &mesh.vertices()[cast_usize(triangle_face.vertices.0)];
+                    let point_b = &mesh.vertices()[cast_usize(triangle_face.vertices.1)];
+                    let point_c = &mesh.vertices()[cast_usize(triangle_face.vertices.2)];
+                    // Calculate the density of points on the respective face
+                    let ab_distance_sq = nalgebra::distance_squared(point_a, point_b);
+                    let bc_distance_sq = nalgebra::distance_squared(point_b, point_c);
+                    let ca_distance_sq = nalgebra::distance_squared(point_c, point_a);
+                    let longest_edge_len = ab_distance_sq
+                        .max(bc_distance_sq.max(ca_distance_sq))
+                        .sqrt();
+                    // Number of face divisions (points) in each direction
+                    let divisions = (longest_edge_len / shortest_voxel_dimension).ceil() as usize;
+                    let divisions_f32 = divisions as f32;
+
+                    for ui in 0..=divisions {
+                        for wi in 0..=divisions {
+                            let u_f32 = ui as f32 / divisions_f32;
+                            let w_f32 = wi as f32 / divisions_f32;
+                            let v_f32 = 1.0 - u_f32 - w_f32;
+                            if v_f32 >= 0.0 {
+                                let barycentric = Point3::new(u_f32, v_f32, w_f32);
+                                // Calculate point position in model space
+                                let cartesian = geometry::barycentric_to_cartesian(
+                                    &barycentric,
+                                    &point_a,
+                                    &point_b,
+                                    &point_c,
+                                );
+                                // and set a voxel containing the point to be on
+                                voxel_cloud.set_cartesian(Voxel::On, &cartesian);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        voxel_cloud
+    }
+
+    /// For each existing voxel turn on all neighbor voxels to grow (offset) the
+    /// volumes stored in the voxel cloud.
+    pub fn grow(&self) -> Self {
```

Comment:
You need to grow the space, otherwise you might (and certainly will in the second iteration) grow out of bounds. Therefore you need to allocate a new chunk of memory, so I don't see any benefits of mutating the existing `VoxelCloud`.

Reply:
Ok, I understand now, thanks!